src/test/resource/test_resource.rs

```diff
@@ -0,0 +1,377 @@
+        for g in SETRLIMITS {
+            println!("  testing fn: {g:?}");
+            // Setting to the same value again should be a no-op and succeed.
+            g.setrlimit(resource.into(), &initial).unwrap();
+            assert_eq!(LibcGetRLimit.getrlimit(resource.into()).unwrap(), initial);
+
+            if initial.rlim_cur < initial.rlim_max {
+                println!("  incrementing");
+                let new_val = linux_api::resource::rlimit {
+                    rlim_cur: initial.rlim_cur + 1,
+                    rlim_max: initial.rlim_max,
+                };
+                g.setrlimit(resource.into(), &new_val).unwrap();
+                assert_eq!(LibcGetRLimit.getrlimit(resource.into()).unwrap(), new_val);
+
+                g.setrlimit(resource.into(), &initial).unwrap();
+            } else {
+                println!("  not incrementing; already at max");
+            }
+
+            if initial.rlim_cur > 0 {
+                println!("  decrementing cur");
+                let new_val = linux_api::resource::rlimit {
+                    rlim_cur: decrement_rlim(initial.rlim_cur),
+                    rlim_max: initial.rlim_max,
+                };
+                g.setrlimit(resource.into(), &new_val).unwrap();
+                assert_eq!(LibcGetRLimit.getrlimit(resource.into()).unwrap(), new_val);
+
+                g.setrlimit(resource.into(), &initial).unwrap();
+            } else {
+                println!("  not decrementing cur; already at 0");
+            }
+
+            // Not tested:
+            // * increasing rlim_max: natively, this only works if we're running
+            // with CAP_SYS_RESOURCE or root; we'd have to either probe for
+            // whether this was the case, or only support running one way or the
+            // other.  Likewise shadow would try to natively increase the
+            // maximum in the handler, which would again depend on permissions.
+            // * decreasing rlim_max: we *could* test this, but wouldn't be able
+            // to reliably restore it again as above. Since we're testing
+            // multiple functions for each resource, we'd have to decrease it
+            // several times without restoring it. Seems too fragile to be
+            // worthwhile.
+        }
+    }
+}
+
+fn test_setrlimits_errors() {
+    for g in SETRLIMITS {
+        println!("  testing fn: {g:?}");
+
+        // Bad resource
+        assert_eq!(
+            g.setrlimit(
+                1234,
+                &linux_api::resource::rlimit {
+                    rlim_cur: 0,
+                    rlim_max: 0
+                }
+            ),
+            Err(Errno::EINVAL)
+        );
+
+        // reducing rlim_max below rlim_cur
+        let cur = LibcGetRLimit
+            .getrlimit(linux_api::resource::Resource::RLIMIT_NOFILE.into())
+            .unwrap();
+        assert!(cur.rlim_cur > 0);
+        assert!(cur.rlim_cur < i64::MAX as u64);
```

Comment:
Maybe skip if the current is >= `i64::MAX`? I wonder if any systems set this as unlimited?